### PR TITLE
Update EPICorrGadget.cpp

### PR DESCRIPTION
**Pull request by pvelasco**
Modified the way we compute the phase difference between navigators for the odd-even phase correction.

It corrects the magnitude of the complex conjugate product of the navigators before combining across coils.

Here is some Matlab code with simulations of different ways of weighting the navigators when computing the phase difference.  The current way of calculating is option 1, while the proposed way is option 2.  All simulations I have ran show that option 2 gives a better fit than any of the other options, both in terms of lower residuals and in terms of a narrower distribution of linear slope estimates.

```
Niter = 10000;   % number of iterations.
Nx = 100;    % number of voxels along RO direction
Nc = 6;      % number of coils
a  = 0.1;    % actual phase difference from one point to the next
n  = 0.3;    % noise-to-signal ratio (e.g. 0.1 means noise std is 1/10th of signal)

rho = sin( pi/Nx*(0:Nx-1)' );  % something simple, we just want some not-constant profile
n1  = rho;                     % let's make it simple: phase of nav1 = 0
n2  = rho.*exp( 1i*a*(0:Nx-1)');

x = linspace(-0.5,0.5,Nx)';    % position along RO direction


% initialize variables:
slope_1 = zeros(Niter,1);
slope_2 = zeros(Niter,1);
slope_3 = zeros(Niter,1);
slope_4 = zeros(Niter,1);
slope_5 = zeros(Niter,1);
std_res_ph_1 = zeros(Niter,1);
std_res_ph_2 = zeros(Niter,1);
std_res_ph_3 = zeros(Niter,1);
std_res_ph_4 = zeros(Niter,1);
std_res_ph_5 = zeros(Niter,1);

for it = 1:Niter
  % add some noise to the signal:
  n1_i = repmat(real(n1),1,Nc)+n*randn(Nx,Nc) + 1i*( repmat(imag(n1),1,Nc)+n*randn(Nx,Nc) );
  n2_i = repmat(real(n2),1,Nc)+n*randn(Nx,Nc) + 1i*( repmat(imag(n2),1,Nc)+n*randn(Nx,Nc) );
  
  z = conj(n1_i).*n2_i;  % simulated complex data we want to fit
  
  %% option 1 (current estimate):
  % add across coils, compute phase difference, & sum along RO:
  z_mean = sum( z, 2 );
  z_mean_diff = conj(z_mean(1:end-1)).*z_mean(2:end);
  slope_1(it) = (Nx-1) * angle( sum( z_mean_diff ) );

  %% option 2 (correction first thing):
  % first correct for the sqrt, then add across coils, compute
  % phase difference, and sum along RO:
  z_corr = z./sqrt(abs(z));
  z_corr_mean = sum( z_corr, 2 );
  z_corr_mean_diff = conj(z_corr_mean(1:end-1)).*z_corr_mean(2:end);
  slope_2(it) = (Nx-1) * angle( sum( z_corr_mean_diff ) );

  %% option 3 (correction after coil average):
  % first add across coils, then correct for the sqrt, compute
  % phase difference, and sum along RO:
  z_mean_corr = z_mean ./ sqrt(abs( z_mean ));
  z_mean_corr_diff = conj(z_mean_corr(1:end-1)).*z_mean_corr(2:end);
  slope_3(it) = (Nx-1) * angle( sum( z_mean_corr_diff ) );

  %% option 4 (correction after computing phase diff along RO):
  % first add across coils, then compute phase difference,
  % correct for the sqrt, and sum along RO:
  z_mean_diff_corr = z_mean_diff ./ sqrt(abs(z_mean_diff));
  slope_4(it) = (Nx-1) * angle( sum( z_mean_diff_corr ) );

  %% option 5 (correction after each time we multiply signals):
  % first correct for the sqrt, then add across coils, compute
  % phase difference, correct again, and sum along RO:
  z_corr_mean_diff_corr = z_corr_mean_diff ./ sqrt(abs(z_corr_mean_diff));
  slope_5(it) = (Nx-1) * angle( sum( z_corr_mean_diff_corr ) );
  
  %% now, calculate the intercepts:
  z_corr_1 = z .* repmat(exp( -1i*slope_1(it)*x ),1,Nc);
  z_corr_2 = z .* repmat(exp( -1i*slope_2(it)*x ),1,Nc);
  z_corr_3 = z .* repmat(exp( -1i*slope_3(it)*x ),1,Nc);
  z_corr_4 = z .* repmat(exp( -1i*slope_4(it)*x ),1,Nc);
  z_corr_5 = z .* repmat(exp( -1i*slope_5(it)*x ),1,Nc);
  
  intercept_1 = angle(sum( z_corr_1(:) ));      % average across coils and RO
  intercept_2 = angle(sum( z_corr_2(:)./sqrt(abs(z_corr_2(:))) ));   % correct first
  z_corr_3_mean = sum( z_corr_3, 2 );
  intercept_3 = angle(sum( z_corr_3_mean./sqrt(abs(z_corr_3_mean)) ));
  intercept_4 = angle(sum( z_corr_4(:) ));
  intercept_5 = angle(sum( z_corr_5(:)./sqrt(abs(z_corr_5(:))) ));

  
  %% and the corrected z:
  z_corrected_1 = z .* repmat(exp( -1i*(slope_1(it)*x+intercept_1) ),1,Nc);
  z_corrected_2 = z .* repmat(exp( -1i*(slope_2(it)*x+intercept_2) ),1,Nc);
  z_corrected_3 = z .* repmat(exp( -1i*(slope_3(it)*x+intercept_3) ),1,Nc);
  z_corrected_4 = z .* repmat(exp( -1i*(slope_4(it)*x+intercept_4) ),1,Nc);
  z_corrected_5 = z .* repmat(exp( -1i*(slope_5(it)*x+intercept_5) ),1,Nc);
  
  % store the mean of the residual phase:
  std_res_ph_1(it) = mean(abs(angle(z_corrected_1(:))).^2);
  std_res_ph_2(it) = mean(abs(angle(z_corrected_2(:))).^2);
  std_res_ph_3(it) = mean(abs(angle(z_corrected_3(:))).^2);
  std_res_ph_4(it) = mean(abs(angle(z_corrected_4(:))).^2);
  std_res_ph_5(it) = mean(abs(angle(z_corrected_5(:))).^2);

end
```
The best fit would have the lowest residual phase, so compare all the std we have computed for both cases (we compare pairwise):
```
plot_width = 0.21; plot_height = 0.21;
offset_x   = 0.1; offset_y    = 0.1;
figure(1)
clf
plot_min = min(min([std_res_ph_1 std_res_ph_2 std_res_ph_3 std_res_ph_4 std_res_ph_5]));
plot_max = max(max([std_res_ph_1 std_res_ph_2 std_res_ph_3 std_res_ph_4 std_res_ph_5]));
my_lim = [plot_min 0.9*plot_max];
for x = 1:5
  for y = x+1:5
    axes('Position',[offset_x+(x-1)*plot_width,...
                     offset_y+(y-2)*plot_height,...
                     plot_width, plot_height])
    my_str = sprintf('plot(std_res_ph_%d, std_res_ph_%d,''.'')',x,y);
    eval(my_str);
    % plot the identity:
    hold on
    plot(my_lim,my_lim,'r')
    xlim(my_lim);
    ylim(my_lim);
    if x==1
      ylabel(sprintf('option %d',y));
    else
      set(gca,'YTickLabel','')
    end
    if y==x+1
      xlabel(sprintf('option %d',x));
    else
      set(gca,'XTickLabel','')
    end
    if x==2 && y==5
      title('std of the residual phase')
    end
  end
end
```

phase--linear_fit_sqrt_normalization

To compare between options, the figure above shows the standard deviation of the residual phase after the linear phase removal, with each point in the distribution representing one iteration.  For each comparison plot, the distribution will be skewed towards the axis with the higher standard deviation of each pair, so the other one is better.  Looking at the top row, we can see option 5 has higher residuals than all other options, so option 5 is the worst and should be excluded.  Next, looking at the left column, we can see how option 1 has higher residuals than options 2--4, so it should be excluded.  Moving then to the second column, one can see that option 2 is better than both options 3--5.  (Also, note that options 3 and 4 are identical.  It makes sense, because taking the square root and multiplying commute.

The best fit would give a more accurate (less bias) estimate of the slope, with a narrower distribution. (I'm not including option 4 since we find it equivalent to option 3):
```
figure(2)
% Get the histogram values:
[counts,centers] = hist([slope_1 slope_2 slope_3 slope_5]/(Nx-1),50);
% Now, plot the distributions (as lines), for clarity:
clf
plot(centers, counts)
xlabel('slope estimate')
legend(sprintf('option 1\n(mean: %6.5f)',mean(slope_1)/(Nx-1)),...
       sprintf('option 2\n(mean: %6.5f)',mean(slope_2)/(Nx-1)),...
       sprintf('option 3\n(mean: %6.5f)',mean(slope_3)/(Nx-1)),...
       sprintf('option 5\n(mean: %6.5f)',mean(slope_5)/(Nx-1)) )
% You can change the x-axis, to have more detail:
xlim([centers(7) centers(end-6)])
```
[phase--linear_fit_sqrt_normalization_histogram.pdf](https://github.com/gadgetron/gadgetron/files/501444/phase--linear_fit_sqrt_normalization_histogram.pdf)

This figure shows the distribution of slope values obtained for the different options.  Note that option 2 has the narrowest distribution.  This is corroborated when calculating the standard deviation of obtained values:

`std([slope_1 slope_2 slope_3 slope_5]/(Nx-1))
`
The smallest value is for option 2.

### CODE REVIEW

**Comment by pvelasco** [comment]
Please note that this new version doesn't pass the epi_2d integration test.  If everybody agrees with the proposed changes, I will update the reference file for the epi_2d integration test.

**Comment by xueh2** [comment]
I read it through. The results are very solid.  We can update integration test cases ?

**Comment by pvelasco** [comment]
Hi Hui @xueh2,
Thanks for taking the time to read through it.

I have updated the image (epi_2d_out_20160929_pjv.h5) to:
https://drive.google.com/file/d/0B3Wx_3B2y_lZUzZtQ253d2dlLXM/view?usp=sharing
Please let me know if you have problems accessing it.

I will update data.txt and epi_2d.cfg files and add them to the commit.

**Comment by hansenms** [comment]
I have uploaded the file and will merge.

**Comment by pvelasco** [comment]
Hi @hansenms and @xueh2,
I just realized that I made a mistake: what we care about is minimizing the residual of the phase **where there is signal**, not just everywhere.  So when I compared the residuals of the different options I should have looked at a weighted average.
When I do that (I'm attaching the code and figures in the next comment), I find that option 1 is the preferred one, not option 2.  Option 1 was the existing code before this PR.
What should we do?  Is there a way to revert to the previous version, or should I do another PR reverting the changes?
Sorry for the mess...

**Comment by pvelasco** [comment]
Here is the correct Matlab code, which checks the **weighted residual phases** for the different options:

'''
Niter = 10000;   % number of iterations.
Nx = 100;    % number of voxels along RO direction
Nc = 6;      % number of coils
a  = 0.1;    % actual phase diff from one point to the next
n  = 0.01;    % noise-to-signal ratio (e.g. 0.1 means noise std is 1/10th of signal)

rho = -cos( 2*pi/Nx*(0:Nx-1)' );  % something simple, we just want some not-constant profile
rho(rho<0) = 0;							% ...but with some background in it
n1  = rho;                     % let's make it simple: phase of nav1 = 0
n2  = rho.*exp( 1i*a*(0:Nx-1)');

x = linspace(-0.5,0.5,Nx)';    % position along RO direction


% initialize variables:
slope_1 = zeros(Niter,1);
slope_2 = zeros(Niter,1);
slope_3 = zeros(Niter,1);
slope_4 = zeros(Niter,1);
slope_5 = zeros(Niter,1);
std_res_ph_1 = zeros(Niter,1);
std_res_ph_2 = zeros(Niter,1);
std_res_ph_3 = zeros(Niter,1);
std_res_ph_4 = zeros(Niter,1);
std_res_ph_5 = zeros(Niter,1);

for it = 1:Niter
  % add some noise to the signal:
  n1_i = repmat(real(n1),1,Nc)+n*randn(Nx,Nc) + 1i*( repmat(imag(n1),1,Nc)+n*randn(Nx,Nc) );
  n2_i = repmat(real(n2),1,Nc)+n*randn(Nx,Nc) + 1i*( repmat(imag(n2),1,Nc)+n*randn(Nx,Nc) );
  
  z = conj(n1_i).*n2_i;  % simulated complex data we want to fit
  
  %% option 1 (current estimate):
  % add across coils, compute phase difference, & sum along RO:
  z_mean = sum( z, 2 );
  z_mean_diff = conj(z_mean(1:end-1)).*z_mean(2:end);
  slope_1(it) = (Nx-1) * angle( sum( z_mean_diff ) );

  %% option 2 (correction first thing):
  % first correct for the sqrt, then add across coils, compute
  % phase difference, and sum along RO:
  z_corr = z./sqrt(abs(z));
  z_corr_mean = sum( z_corr, 2 );
  z_corr_mean_diff = conj(z_corr_mean(1:end-1)).*z_corr_mean(2:end);
  slope_2(it) = (Nx-1) * angle( sum( z_corr_mean_diff ) );

  %% option 3 (correction after coil average):
  % first add across coils, then correct for the sqrt, compute
  % phase difference, and sum along RO:
  z_mean_corr = z_mean ./ sqrt(abs( z_mean ));
  z_mean_corr_diff = conj(z_mean_corr(1:end-1)).*z_mean_corr(2:end);
  slope_3(it) = (Nx-1) * angle( sum( z_mean_corr_diff ) );

  %% option 4 (correction after computing phase diff along RO):
  % first add across coils, then compute phase difference,
  % correct for the sqrt, and sum along RO:
  z_mean_diff_corr = z_mean_diff ./ sqrt(abs(z_mean_diff));
  slope_4(it) = (Nx-1) * angle( sum( z_mean_diff_corr ) );

  %% option 5 (correction after each time we multiply signals):
  % first correct for the sqrt, then add across coils, compute
  % phase difference, correct again, and sum along RO:
  z_corr_mean_diff_corr = z_corr_mean_diff ./ sqrt(abs(z_corr_mean_diff));
  slope_5(it) = (Nx-1) * angle( sum( z_corr_mean_diff_corr ) );
  
  %% now, calculate the intercepts:
  z_corr_1 = z .* repmat(exp( -1i*slope_1(it)*x ),1,Nc);
  z_corr_2 = z .* repmat(exp( -1i*slope_2(it)*x ),1,Nc);
  z_corr_3 = z .* repmat(exp( -1i*slope_3(it)*x ),1,Nc);
  z_corr_4 = z .* repmat(exp( -1i*slope_4(it)*x ),1,Nc);
  z_corr_5 = z .* repmat(exp( -1i*slope_5(it)*x ),1,Nc);
  
  intercept_1 = angle(sum( z_corr_1(:) ));      % average across coils and RO
  intercept_2 = angle(sum( z_corr_2(:)./sqrt(abs(z_corr_2(:))) ));   % correct first
  z_corr_3_mean = sum( z_corr_3, 2 );
  intercept_3 = angle(sum( z_corr_3_mean./sqrt(abs(z_corr_3_mean)) ));
  intercept_4 = angle(sum( z_corr_4(:) ));
  intercept_5 = angle(sum( z_corr_5(:)./sqrt(abs(z_corr_5(:))) ));

  
  %% and the corrected z:
  z_corrected_1 = z .* repmat(exp( -1i*(slope_1(it)*x+intercept_1) ),1,Nc);
  z_corrected_2 = z .* repmat(exp( -1i*(slope_2(it)*x+intercept_2) ),1,Nc);
  z_corrected_3 = z .* repmat(exp( -1i*(slope_3(it)*x+intercept_3) ),1,Nc);
  z_corrected_4 = z .* repmat(exp( -1i*(slope_4(it)*x+intercept_4) ),1,Nc);
  z_corrected_5 = z .* repmat(exp( -1i*(slope_5(it)*x+intercept_5) ),1,Nc);
  
  % store the mean of the weighted residual phase:
  std_res_ph_1(it) = mean(abs(abs(z_corrected_1(:)).*angle(z_corrected_1(:))).^2);
  std_res_ph_2(it) = mean(abs(abs(z_corrected_2(:)).*angle(z_corrected_2(:))).^2);
  std_res_ph_3(it) = mean(abs(abs(z_corrected_3(:)).*angle(z_corrected_3(:))).^2);
  std_res_ph_4(it) = mean(abs(abs(z_corrected_4(:)).*angle(z_corrected_4(:))).^2);
  std_res_ph_5(it) = mean(abs(abs(z_corrected_5(:)).*angle(z_corrected_5(:))).^2);
end
'''
The best fit would have the lowest residual phase, so compare all the std we have computed for both cases (we compare pairwise):

'''
plot_width = 0.21; plot_height = 0.21;
offset_x   = 0.1; offset_y    = 0.1;
figure(1)
clf
% Calculate the plot limits.
% (We don't use option 5 because the residuals are much higher)
plot_min = min(min([std_res_ph_1 std_res_ph_2 std_res_ph_3 std_res_ph_4]));
plot_max = max(max([std_res_ph_1 std_res_ph_2 std_res_ph_3 std_res_ph_4]));
my_lim = [plot_min 0.9*plot_max];
for x = 1:5
  for y = x+1:5
    axes('Position',[offset_x+(x-1)*plot_width,...
                     offset_y+(y-2)*plot_height,...
                     plot_width, plot_height])
    my_str = sprintf('plot(std_res_ph_%d, std_res_ph_%d,''.'')',x,y);
    eval(my_str);
    % plot the identity:
    hold on
    plot(my_lim,my_lim,'r')
    xlim(my_lim);
    ylim(my_lim);
    if x==1
      ylabel(sprintf('option %d',y));
    else
      set(gca,'YTickLabel','')
    end
    if y==x+1
      xlabel(sprintf('option %d',x));
    else
      set(gca,'XTickLabel','')
    end
    if x==2 && y==5
      title('std of the weighted residual phase')
    end
  end
end
'''
phase--linear_fit_sqrt_normalization

To compare between options, the figure above shows the standard deviation of the weighted residual phase after the linear phase removal (the weights are given by the magnitude of the vector we want to fit), with each point in the distribution representing one iteration.  For each comparison plot, the distribution will be skewed towards the axis with the higher standard deviation of each pair, so the other one is better.  Looking at the top row, we can see option 5 has higher residuals than all other options, so option 5 is the worst and should be excluded.  Next, looking at the left-most column, we can see how option 1 has lower residuals than all the others.

The best fit would give a more accurate (less bias) estimate of the slope, with a narrower distribution. (I'm not including option 4 since we find it equivalent to option 3):

'''
figure(2)
% Get the histogram values:
[counts,centers] = hist([slope_1 slope_2 slope_3 slope_5]/(Nx-1),50);
% Now, plot the distributions (as lines), for clarity:
clf
plot(centers, counts)
xlabel('slope estimate')
legend(sprintf('option 1\n(std: %6.5f)',std(slope_1)/(Nx-1)),...
       sprintf('option 2\n(std: %6.5f)',std(slope_2)/(Nx-1)),...
       sprintf('option 3\n(std: %6.5f)',std(slope_3)/(Nx-1)),...
       sprintf('option 5\n(std: %6.5f)',std(slope_5)/(Nx-1)) )
% You can change the x-axis, to have more detail:
xlim([centers(7) centers(end-6)])
'''
[phase--linear_fit_sqrt_normalization_histogram.pdf](https://github.com/gadgetron/gadgetron/files/506744/phase--linear_fit_sqrt_normalization_histogram.pdf)

This figure shows the distribution of slope values obtained for the different options. Note that option 1 has the narrowest distribution (I'm giving the std of the slope estimates for the different options in the figure's legend). This is corroborated when calculating the standard deviation of obtained values:

std([slope_1 slope_2 slope_3 slope_4 slope_5]/(Nx-1)) 
The smallest value is for option 1.

(Note that when the SNR is very low (e.g. n=0.3), the std of the slope distribution is narrower for option 2, but the residual comparison is still favorable to option 1.)

Bottom line: option 1 is the preferred one.

**Comment by hansenms** [comment]
Please do a PR to revert the changes.
